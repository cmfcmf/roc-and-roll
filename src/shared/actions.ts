import { createAction } from "@reduxjs/toolkit";
import type { Update as OriginalUpdate } from "@reduxjs/toolkit";
import {
  EphemeralPlayer,
  InitiativeTrackerSyncedState,
  RRID,
  RRInitiativeTrackerEntry,
  RRInitiativeTrackerEntryLairAction,
  RRInitiativeTrackerEntryCharacter,
  RRLogEntry,
  RRLogEntryDiceRoll,
  RRLogEntryMessage,
  RRMap,
  RRMapObject,
  RRMapID,
  RRPlayer,
  RRPlayerID,
  RRPrivateChat,
  RRPrivateChatID,
  RRPrivateChatMessage,
  RRCharacter,
  RRCharacterID,
  RRLogEntryAchievement,
  RRActiveSongOrSoundSet,
  RRAssetImage,
  RRAssetSong,
  RRToken,
  RRMapObjectID,
  RRAssetID,
  RRGlobalSettings,
  EMPTY_ENTITY_COLLECTION,
  RRPlaylist,
  RRPlaylistEntryID,
  RRSoundSet,
  RRPlaylistID,
  RRPlaylistEntry,
  RRAsset,
  RRDiceTemplateCategoryID,
  RRDiceTemplateID,
  RRDiceTemplatePart,
  RRDiceTemplatePartID,
} from "./state";
import { rrid, timestamp } from "./util";
import type {
  RRDiceTemplate,
  // We need this import for #reasons, otherwise TypeScript won't compile.
  // eslint-disable-next-line unused-imports/no-unused-imports
  _RRDiceTemplate,
  RRDiceTemplateCategory,
} from "./validation";

type Update<T extends { id: RRID }> = OriginalUpdate<T, T["id"]>;

// Omits a property from a union of types
//
// CC-BY-SA 4.0 by jcalz
// https://stackoverflow.com/a/57103940/2560557
// https://stackoverflow.com/users/2887218/jcalz
type DistributiveOmit<T, K extends keyof T> = T extends unknown
  ? Omit<T, K>
  : never;

////////////////////////////////////////////////////////////////////////////////
// Players
////////////////////////////////////////////////////////////////////////////////

export const playerAdd = createAction(
  "player/add",
  (player: Omit<RRPlayer, "id">): { payload: RRPlayer } => ({
    payload: { id: rrid<RRPlayer>(), ...player },
  })
);

export const playerUpdate = createAction<Update<RRPlayer>>("player/update");

export const playerUpdateAddFavoriteAssetId = createAction<{
  id: RRPlayerID;
  assetId: RRAssetID;
}>("player/update/assetId/add");

export const playerUpdateRemoveFavoriteAssetId = createAction<{
  id: RRPlayerID;
  assetId: RRAssetID;
}>("player/update/assetId/remove");

export const playerRemove = createAction<RRPlayer["id"]>("player/remove");

////////////////////////////////////////////////////////////////////////////////
// Characters
////////////////////////////////////////////////////////////////////////////////

export const characterAdd = createAction(
  "character/add",
  (character: Omit<RRCharacter, "id">): { payload: RRCharacter } => ({
    payload: { id: rrid<RRCharacter>(), ...character },
  })
);

export const characterUpdate =
  createAction<Update<RRCharacter>>("character/update");

export const characterRemove =
  createAction<RRCharacter["id"]>("character/remove");

export const playerUpdateAddCharacterId = createAction<{
  id: RRPlayerID;
  characterId: RRCharacterID;
}>("player/update/characterId");

export const characterAddDiceTemplateCategory = createAction<{
  id: RRCharacterID;
  category: RRDiceTemplateCategory;
}>("character/diceTemplateCategories/add");

export const characterUpdateDiceTemplateCategory = createAction<{
  id: RRCharacterID;
  category: Update<RRDiceTemplateCategory>;
}>("character/diceTemplateCategories/update");

export const characterDeleteDiceTemplateCategory = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
}>("character/diceTemplateCategories/delete");

export const characterRemoveDiceTemplate = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
  templateId: RRDiceTemplateID;
}>("character/diceTemplateCategories/template/remove");

export const characterUpdateDiceTemplate = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
  template: Update<RRDiceTemplate>;
}>("character/diceTemplateCategories/template/update");

export const characterUpdateDiceTemplatePart = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
  templateId: RRDiceTemplateID;
  part: Update<RRDiceTemplatePart>;
}>("character/diceTemplateCategories/template/part/update");

export const characterAddDiceTemplatePart = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
  templateId: RRDiceTemplateID;
  part: RRDiceTemplatePart;
}>("character/diceTemplateCategories/template/part/add");

export const characterRemoveDiceTemplatePart = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
  templateId: RRDiceTemplateID;
  partId: RRDiceTemplatePartID;
}>("character/diceTemplateCategories/template/part/remove");

export const characterAddDiceTemplate = createAction<{
  id: RRCharacterID;
  categoryId: RRDiceTemplateCategoryID;
  template: RRDiceTemplate;
}>("character/diceTemplateCategories/template/add");

////////////////////////////////////////////////////////////////////////////////
// Maps
////////////////////////////////////////////////////////////////////////////////

export const mapAdd = createAction("map/add", (map: Omit<RRMap, "id">) => ({
  payload: { id: rrid<RRMap>(), ...map },
}));

export const mapUpdate = createAction<Update<RRMap>>("map/update");

export const mapRemove = createAction<RRMap["id"]>("map/remove");

export const mapSettingsUpdate = createAction<{
  id: RRMapID;
  changes: Partial<RRMap["settings"]>;
}>("map/settings/update");

export const mapObjectAdd = createAction(
  "map/object/add",
  (
    mapId: RRMapID,
    mapObject: DistributiveOmit<RRMapObject, "id">
  ): {
    payload: { mapId: RRMapID; mapObject: RRMapObject };
  } => {
    return {
      payload: {
        mapId,
        mapObject: {
          ...mapObject,
          id: rrid<RRMapObject>(),
        },
      },
    };
  }
);

export const mapObjectUpdate = createAction(
  "map/object/update",
  (mapId: RRMapID, update: Update<RRMapObject>) => {
    return {
      payload: {
        mapId,
        update,
      },
    };
  }
);

interface TokenRemoval {
  mapId: RRMapID;
  mapObject: RRToken;
  relatedCharacter: RRCharacter;
}
interface OtherMapObjectRemoval {
  mapId: RRMapID;
  mapObjectId: RRMapObjectID;
}

function isTokenRemoval(
  o: TokenRemoval | OtherMapObjectRemoval
): o is TokenRemoval {
  return "mapObject" in o;
}

export const mapObjectRemove = createAction(
  "map/object/remove",
  (removal: TokenRemoval | OtherMapObjectRemoval) => {
    return {
      payload: {
        mapId: removal.mapId,
        mapObjectId: isTokenRemoval(removal)
          ? removal.mapObject.id
          : removal.mapObjectId,
        removeTemplateId:
          isTokenRemoval(removal) && removal.relatedCharacter.localToMap
            ? removal.relatedCharacter.id
            : null,
      },
    };
  }
);

////////////////////////////////////////////////////////////////////////////////
// PrivateChats
////////////////////////////////////////////////////////////////////////////////

export const privateChatAdd = createAction(
  "privateChat/add",
  (
    player1Id: RRPlayerID,
    player2Id: RRPlayerID
  ): { payload: RRPrivateChat } => {
    let idA, idB;
    if (player1Id < player2Id) {
      idA = player1Id;
      idB = player2Id;
    } else {
      idA = player2Id;
      idB = player1Id;
    }

    return {
      payload: {
        id: rrid<RRPrivateChat>(),
        idA,
        idB,
        messages: EMPTY_ENTITY_COLLECTION,
      },
    };
  }
);

export const privateChatUpdate =
  createAction<Update<Omit<RRPrivateChat, "messages">>>("privateChat/update");

export const privateChatRemove =
  createAction<RRPrivateChat["id"]>("privateChat/remove");

export const privateChatMessageAdd = createAction(
  "privateChat/message/add",
  (
    chatId: RRPrivateChatID,
    message: Omit<RRPrivateChatMessage, "id" | "timestamp" | "read">
  ): {
    payload: { chatId: RRPrivateChatID; message: RRPrivateChatMessage };
  } => {
    return {
      payload: {
        chatId,
        message: {
          id: rrid<RRPrivateChatMessage>(),
          timestamp: timestamp(),
          read: false,
          ...message,
        },
      },
    };
  }
);

export const privateChatMessageUpdate = createAction(
  "privateChat/message/update",
  (
    chatId: RRPrivateChatID,
    update: Update<Omit<RRPrivateChatMessage, "timestamp">>
  ) => {
    return {
      payload: {
        chatId,
        update,
      },
    };
  }
);

////////////////////////////////////////////////////////////////////////////////
// LogEntries
////////////////////////////////////////////////////////////////////////////////

export const logEntryMessageAdd = createAction(
  "logEntry/message/add",
  (
    logEntry: Omit<RRLogEntryMessage, "id" | "type" | "timestamp">
  ): { payload: RRLogEntryMessage } => ({
    payload: {
      id: rrid<RRLogEntryMessage>(),
      type: "message",
      timestamp: timestamp(),
      ...logEntry,
    },
  })
);

export const logEntryAchievementAdd = createAction(
  "logEntry/achievement/add",
  (
    logEntry: Omit<RRLogEntryAchievement, "id" | "type" | "timestamp">
  ): { payload: RRLogEntryAchievement } => ({
    payload: {
      id: rrid<RRLogEntryAchievement>(),
      type: "achievement",
      timestamp: timestamp(),
      ...logEntry,
    },
  })
);

export const logEntryDiceRollAdd = createAction(
  "logEntry/diceRoll/add",
  (
    logEntry: Omit<RRLogEntryDiceRoll, "id" | "type" | "timestamp">
  ): { payload: RRLogEntryDiceRoll } => ({
    payload: {
      id: rrid<RRLogEntryDiceRoll>(),
      type: "diceRoll",
      timestamp: timestamp(),
      ...logEntry,
    },
  })
);

export const logEntryRemove = createAction<RRLogEntry["id"]>("logEntry/remove");

////////////////////////////////////////////////////////////////////////////////
// InitiativeTracker
////////////////////////////////////////////////////////////////////////////////

export const initiativeTrackerSetVisible = createAction<
  InitiativeTrackerSyncedState["visible"]
>("initiativeTracker/visible");

export const initiativeTrackerSetCurrentEntry = createAction<
  InitiativeTrackerSyncedState["currentEntryId"]
>("initiativeTracker/currentEntryId");

export const initiativeTrackerEntryCharacterAdd = createAction(
  "initiativeTracker/entry/character/add",
  (
    initiativeTrackerEntry: Omit<
      RRInitiativeTrackerEntryCharacter,
      "id" | "type"
    >
  ): { payload: RRInitiativeTrackerEntryCharacter } => ({
    payload: {
      id: rrid<RRInitiativeTrackerEntryCharacter>(),
      type: "character",
      ...initiativeTrackerEntry,
    },
  })
);

export const initiativeTrackerEntryLairActionAdd = createAction(
  "initiativeTracker/entry/lairAction/add",
  (
    initiativeTrackerEntry: Omit<
      RRInitiativeTrackerEntryLairAction,
      "id" | "type"
    >
  ): { payload: RRInitiativeTrackerEntryLairAction } => ({
    payload: {
      id: rrid<RRInitiativeTrackerEntryLairAction>(),
      type: "lairAction",
      ...initiativeTrackerEntry,
    },
  })
);

export const initiativeTrackerEntryCharacterUpdate = createAction<
  Update<RRInitiativeTrackerEntryCharacter>
>("initiativetracker/entry/character/update");

export const initiativeTrackerEntryLairActionUpdate = createAction<
  Update<RRInitiativeTrackerEntryLairAction>
>("initiativetracker/entry/lairaction/update");

export const initiativeTrackerEntryRemove = createAction<
  RRInitiativeTrackerEntry["id"]
>("initiativeTracker/entry/remove");

////////////////////////////////////////////////////////////////////////////////
// Ephemeral state
////////////////////////////////////////////////////////////////////////////////

export const ephemeralPlayerAdd = createAction<EphemeralPlayer>(
  "ephemeral/player/add"
);

export const ephemeralPlayerUpdate = createAction<Update<EphemeralPlayer>>(
  "ephemeral/player/update"
);

export const ephemeralPlayerRemove = createAction<EphemeralPlayer["id"]>(
  "ephemeral/player/remove"
);

export const ephemeralMusicAdd = createAction<RRActiveSongOrSoundSet>(
  "ephemeral/music/add"
);

export const ephemeralMusicUpdate = createAction<
  Update<RRActiveSongOrSoundSet>
>("ephemeral/music/update");

export const ephemeralMusicRemove = createAction<RRActiveSongOrSoundSet["id"]>(
  "ephemeral/music/remove"
);

////////////////////////////////////////////////////////////////////////////////
// Asset state
////////////////////////////////////////////////////////////////////////////////

export const assetSongAdd = createAction(
  "asset/song/add",
  (asset: Omit<RRAssetSong, "id">): { payload: RRAssetSong } => ({
    payload: { id: rrid<RRAssetSong>(), ...asset },
  })
);

export const assetImageAdd = createAction(
  "asset/image/add",
  (asset: Omit<RRAssetImage, "id">): { payload: RRAssetImage } => ({
    payload: { id: rrid<RRAssetImage>(), ...asset },
  })
);

export const assetSongUpdate =
  createAction<Update<RRAssetSong>>("asset/song/update");

export const assetImageUpdate =
  createAction<Update<RRAssetImage>>("asset/image/update");

export const assetRemove = createAction<RRAsset["id"]>("asset/remove");

////////////////////////////////////////////////////////////////////////////////
// Sound Sets
////////////////////////////////////////////////////////////////////////////////

export const soundSetAdd = createAction(
  "soundSet/add",
  (soundSet: Omit<RRSoundSet, "id">): { payload: RRSoundSet } => ({
    payload: { id: rrid<RRSoundSet>(), ...soundSet },
  })
);

export const soundSetUpdate =
  createAction<Update<RRSoundSet>>("soundSet/update");

export const soundSetRemove = createAction<RRSoundSet["id"]>("soundSet/remove");

export const soundSetPlaylistAdd = createAction(
  "soundSet/playlist/add",
  (
    soundSetId: RRSoundSet["id"],
    playlist: Omit<RRPlaylist, "id">
  ): { payload: { soundSetId: RRSoundSet["id"]; playlist: RRPlaylist } } => ({
    payload: { soundSetId, playlist: { id: rrid<RRPlaylist>(), ...playlist } },
  })
);

export const soundSetPlaylistUpdate = createAction<{
  soundSetId: RRSoundSet["id"];
  update: Update<RRPlaylist>;
}>("soundSet/playlist/update");

export const soundSetPlaylistRemove = createAction<{
  soundSetId: RRSoundSet["id"];
  playlistId: RRPlaylistID;
}>("soundSet/playlist/remove");

export const soundSetPlaylistEntryAdd = createAction(
  "soundSet/playlist/entry/add",
  (
    soundSetId: RRSoundSet["id"],
    playlistId: RRPlaylistID,
    playlistEntry: DistributiveOmit<RRPlaylistEntry, "id">
  ): {
    payload: {
      soundSetId: RRSoundSet["id"];
      playlistId: RRPlaylistID;
      playlistEntry: RRPlaylistEntry;
    };
  } => ({
    payload: {
      soundSetId,
      playlistId,
      playlistEntry: { id: rrid<RRPlaylistEntry>(), ...playlistEntry },
    },
  })
);

export const soundSetPlaylistEntryUpdate = createAction<{
  soundSetId: RRSoundSet["id"];
  playlistId: RRPlaylistID;
  update: Update<RRPlaylistEntry>;
}>("soundSet/playlist/entry/update");

export const soundSetPlaylistEntryMove = createAction<{
  soundSetId: RRSoundSet["id"];
  playlistId: RRPlaylistID;
  playlistEntryId: RRPlaylistEntryID;
  direction: "up" | "down";
}>("soundSet/playlist/entry/move");

export const soundSetPlaylistEntryRemove = createAction<{
  soundSetId: RRSoundSet["id"];
  playlistId: RRPlaylist["id"];
  playlistEntryId: RRPlaylistEntryID;
}>("soundSet/playlist/entry/remove");

////////////////////////////////////////////////////////////////////////////////
// Global settings
////////////////////////////////////////////////////////////////////////////////

export const globalSettingsUpdate = createAction<Partial<RRGlobalSettings>>(
  "globalSettings/update"
);
