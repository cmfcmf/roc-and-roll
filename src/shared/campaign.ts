import { Opaque } from "type-fest";
import { SyncedState } from "./state";

export type CampaignId = Opaque<"gameId", string>;

export type CampaignEntity = {
  readonly id: CampaignId;
  readonly name: string;
};

export type CampaignEntityWithState = CampaignEntity & {
  readonly state: SyncedState;
};
