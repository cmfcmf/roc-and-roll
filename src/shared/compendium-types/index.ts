// NOTE: This folder is located in shared/ because its compilation is slow, not
// because it is actually supposed to be used on the server!
import * as z from "zod";
import { assert, IsExact } from "conditional-type-checks";
import { isRRID } from "../validation";
import { MakeRRID } from "../state";
import { CompendiumSpell, isSpell } from "./spell";
import { CompendiumMonster, isMonster } from "./monster";

export const isCompendiumData = z.strictObject({
  spell: z.array(isSpell),
  monster: z.optional(z.array(isMonster)),
});

export interface CompendiumData {
  spell: CompendiumSpell[];
  monster?: CompendiumMonster[];
}

// Make sure that the schema really matches the CompendiumData type.
assert<IsExact<z.infer<typeof isCompendiumData>, CompendiumData>>(true);

export type CompendiumSourceID = MakeRRID<"compendiumSource">;

export interface CompendiumSource {
  id: CompendiumSourceID;
  title: string;
  meta: string;
  data: CompendiumData;
}

export const isCompendiumSource = z.strictObject({
  id: isRRID<CompendiumSourceID>(),
  title: z.string(),
  meta: z.string(),
  data: isCompendiumData,
});

// Make sure that the schema really matches the TextEntry type.
assert<IsExact<z.infer<typeof isCompendiumSource>, CompendiumSource>>(true);
