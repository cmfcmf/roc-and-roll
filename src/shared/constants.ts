export const GRID_SIZE = 70;

export const USE_CONCURRENT_MODE = true as boolean;

export const TOKEN_SIZES = [0.5, 1, 2, 3, 4, 8, 16].map(
  (size) => size * GRID_SIZE
);

export const SOCKET_IO_PATH = "/api/socket.io";

export const DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME = 500;

export const FORCE_COMMIT_FIELD_VALUE_AFTER = 30 * 1000;

export const STORE_SUBSCRIPTION_THROTTLE = 100;

export const LAST_MIGRATION_VERSION = 12;

export const SYNC_MY_MOUSE_POSITION = true as boolean;

export const DEFAULT_BACKGROUND_IMAGE_HEIGHT = 10 * GRID_SIZE;

export const SOCKET_SET_PLAYER_ID = "SET_PLAYER_ID";

export const SOCKET_SET_STATE = "SET_STATE";

export const SOCKET_PATCH_STATE = "PATCH_STATE";

export const SOCKET_DISPATCH_ACTION = "REDUX_ACTION";

export const SOCKET_BROADCAST_MSG = "MESSAGE";

export const DEFAULT_VOLUME = 0.5;
