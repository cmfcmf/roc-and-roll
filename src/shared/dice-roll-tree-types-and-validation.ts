import * as t from "typanion";
import { assert, IsExact } from "conditional-type-checks";
import { damageTypes } from "./state";

export const isDamageType = t.isObject({
  type: t.isEnum(damageTypes),
});

export type RRDamageType = t.InferType<typeof isDamageType>;

export type DiceRollTree<WithResults extends true | false> =
  DRTPartExpression<WithResults>;

export type DRTPartExpression<WithResults extends true | false> =
  | DRTPartTerm<WithResults>
  | DRTPartParens<WithResults>
  | DRTPartNegated<WithResults>
  | DRTPartDice<WithResults>
  | DRTPartNum;

interface DRTPart<T extends string> {
  type: T;
}

export interface DRTPartTerm<WithResults extends true | false>
  extends DRTPart<"term"> {
  operator: "*" | "/" | "+" | "-";
  operands: DRTPartExpression<WithResults>[];
}

export interface DRTPartParens<WithResults extends true | false>
  extends DRTPart<"parens"> {
  inner: DRTPartExpression<WithResults>;
}

export interface DRTPartDice<WithResults extends true | false>
  extends DRTPart<"dice"> {
  count: number;
  results: WithResults extends true ? number[] : "not-yet-rolled";
  modified: "none" | "advantage" | "disadvantage";
  faces: number;
  damage: RRDamageType;
}

export interface DRTPartNum extends DRTPart<"num"> {
  value: number;
  damage: RRDamageType;
}

export interface DRTPartNegated<WithResults extends true | false>
  extends DRTPart<"negated"> {
  inner: DRTPartExpression<WithResults>;
}

const __isDRTPartExpression = <WithResults extends true | false>(
  withResults: WithResults
) =>
  t.makeValidator({
    test: (value, state): value is DRTPartExpression<WithResults> =>
      isDRTPartExpression(withResults)(value, state),
  });

const isDRTPartParens = <WithResults extends true | false>(
  withResults: WithResults
) =>
  t.isObject({
    type: t.isLiteral("parens"),
    inner: __isDRTPartExpression(withResults),
  });

const isDRTPartNum = t.isObject({
  type: t.isLiteral("num"),
  value: t.isNumber(),
  damage: isDamageType,
});

const isDRTPartDice = <WithResults extends true | false>(
  withResults: WithResults
) =>
  t.applyCascade(
    t.isObject({
      type: t.isLiteral("dice"),
      count: t.isNumber(),
      results: t.makeValidator({
        test: (
          value,
          state
        ): value is WithResults extends true ? number[] : "not-yet-rolled" =>
          withResults
            ? t.isArray(t.isNumber())(value, state)
            : t.isLiteral("not-yet-rolled")(value, state),
      }),
      modified: t.isEnum(["none", "advantage", "disadvantage"] as const),
      faces: t.applyCascade(t.isNumber(), [t.isInteger(), t.isAtLeast(1)]),
      damage: isDamageType,
    }),
    [
      // Validate that count === results.length
      t.makeValidator<Record<string, unknown>>({
        test: (dicePart, state) => {
          if (!withResults) {
            return true;
          }
          if (
            !("results" in dicePart) ||
            !("count" in dicePart) ||
            !Array.isArray(dicePart["results"])
          ) {
            return false;
          }

          if (dicePart["results"].length === dicePart["count"]) {
            return true;
          }

          return t.pushError(
            state,
            `A rolled dicepart should have exactly as many dice roll results as its dice count (count = ${String(
              dicePart["count"]
            )}, results = ${String(dicePart["results"].length)})`
          );
        },
      }),
    ]
  );

const isDRTPartNegated = <WithResults extends true | false>(
  withResults: WithResults
) =>
  t.isObject({
    type: t.isLiteral("negated"),
    inner: __isDRTPartExpression(withResults),
  });

const isDRTPartTerm = <WithResults extends true | false>(
  withResults: WithResults
) =>
  t.isObject({
    type: t.isLiteral("term"),
    operator: t.isEnum(["*", "/", "+", "-"] as const),
    operands: t.applyCascade(t.isArray(__isDRTPartExpression(withResults)), [
      t.hasMinLength(2),
    ]),
  });

const isDRTPartExpression = <WithResults extends true | false>(
  withResults: WithResults
) =>
  t.isOneOf(
    [
      isDRTPartTerm(withResults),
      isDRTPartParens(withResults),
      isDRTPartNegated(withResults),
      isDRTPartDice(withResults),
      isDRTPartNum,
    ],
    { exclusive: true }
  );

export const isDiceRollTree = isDRTPartExpression;

const isDiceRollTreeWithoutResults = isDiceRollTree(false);
const isDiceRollTreeWithResults = isDiceRollTree(true);

// Make sure that the schema really matches the DiceRollTree type.
assert<
  IsExact<t.InferType<typeof isDiceRollTreeWithoutResults>, DiceRollTree<false>>
>(true);
assert<
  IsExact<t.InferType<typeof isDiceRollTreeWithResults>, DiceRollTree<true>>
>(true);
