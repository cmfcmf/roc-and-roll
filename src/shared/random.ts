// Will import just an empty object when used in the browser, because
// resolve.fallback.crypto is set to false in webpack.client.js.
import nodeCrypto from "crypto";

export function randomBetweenInclusive(min: number, max: number) {
  if (max < min) {
    throw new Error(
      `Cannot generate a random number where max is less than min!`
    );
  }
  const range = max - min + 1;

  if (!Number.isInteger(min) || !Number.isInteger(max) || range >= 0xffffffff) {
    throw new Error(`min and/or max are invalid.`);
  }

  const array = new Uint32Array(1);

  if ("randomFillSync" in nodeCrypto) {
    nodeCrypto.randomFillSync(array);
  } else {
    window.crypto.getRandomValues(array);
  }

  return (array[0]! % range) + min;
}
