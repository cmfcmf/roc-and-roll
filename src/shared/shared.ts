export class Foo {
  public x?: string;
}
