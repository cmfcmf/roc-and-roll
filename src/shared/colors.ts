import { randomBetweenInclusive } from "./random";

export const colors = [
  // Generated with https://medialab.github.io/iwanthue/
  "#c49fa6",
  "#a74d63",
  "#dd385d",
  "#df7f81",
  "#b03f36",
  "#e14427",
  "#7e5f56",
  "#e77e5e",
  "#9f5c3c",
  "#ce7225",
  "#d9ae7f",
  "#d9a83c",
  "#836a2a",
  "#d6db33",
  "#8f9c38",
  "#cddb66",
  "#7c855b",
  "#bed994",
  "#78d747",
  "#c9d7c1",
  "#448235",
  "#64d784",
  "#5ca98e",
  "#457264",
  "#6be6c9",
  "#72c4d9",
  "#507691",
  "#5788cf",
  "#bbb8de",
  "#5661d8",
  "#8f7ad0",
  "#8e44de",
  "#765d87",
  "#9445ab",
  "#de96e3",
  "#d64cce",
  "#ba6ba5",
  "#a23372",
  "#e44598",
  "#e68eb1",
];

export function randomColor() {
  return colors[randomBetweenInclusive(0, colors.length - 1)]!;
}
