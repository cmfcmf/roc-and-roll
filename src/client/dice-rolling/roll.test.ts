import { randomBetweenInclusive } from "../../shared/random";
import {
  diceResult,
  diceResultString,
  isValidDiceString,
  parseDiceStringAndRoll,
  RollVisitor,
} from "./roll";

describe("dice rolling", () => {
  beforeEach(() => {
    RollVisitor.rollFunction = (min, max) => {
      return max;
    };
  });

  afterEach(() => {
    RollVisitor.rollFunction = randomBetweenInclusive;
  });

  test.each([
    ["", false],
    ["foo", false],

    ["42", true],
    ["+42", true],
    ["+ 42", true],
    ["1d3", true],
    ["d4", true],
    ["1d20", true],
    ["1a20", true],
    ["1i20", true],
    ["1d20+1", true],
    ["1d20-1", true],
    ["-2d4 - 9", true],
    [" 1d20 + 1 ", true],
    ["1d20 ร 3", true],
    ["1d20 รท 3", true],
  ])('isValidDiceString: "%s" = %s', (diceString, valid) => {
    expect(isValidDiceString(diceString)).toBe(valid);
  });

  it.each([
    ["20 + 10 * 6", 80],
    ["10", 10],
    ["1 + 2", 3],
    ["1 + 2 + 3", 6],
    ["3 * 4", 12],
    ["3 * 4d10 - 1d8", 112],
  ])("calculates the result correctly", (rollString, result) => {
    expect(diceResult(parseDiceStringAndRoll(rollString))).toBe(result);
  });

  it.each([["3 + 3d20 * 10"], ["1+2"]])(
    "displays dice results as string",
    (rollString) => {
      expect(
        diceResultString(parseDiceStringAndRoll(rollString))
      ).toMatchSnapshot();
    }
  );
});
