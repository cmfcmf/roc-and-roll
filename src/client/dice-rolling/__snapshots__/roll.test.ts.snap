// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dice rolling displays dice results as string 1`] = `
<React.Fragment>
  <b>
    3
  </b>
   
  +
   
  <React.Fragment>
    <React.Fragment>
      <b>
        [
        20, 20, 20
        ] (
        3
        d
        20
        )
      </b>
    </React.Fragment>
     
    *
     
    <b>
      10
    </b>
  </React.Fragment>
</React.Fragment>
`;

exports[`dice rolling displays dice results as string 2`] = `
<React.Fragment>
  <b>
    1
  </b>
   
  +
   
  <b>
    2
  </b>
</React.Fragment>
`;
